### PR TITLE
Add `keytype`+`valtype` method for AbstractArray and AbstractVector

### PR DESCRIPTION
**Pull request by tkluck**
The function `keytype(a)` returns `eltype(keys(a))` for associative structures; it should do the same for `AbstractArray` and `AbstractVector`. Looks like this was overlooked(?) in #25577.

#25104 is a related pull request, but renaming is less urgent than adding expected behaviour.

### CODE REVIEW

**Comment by Keno** [comment]
> eltype(keys(a))

Wouldn't literally this definition be a better generic fallback and cover these cases?

**Comment by andyferris** [comment]
I feel the fallback can be allowed to have allocations, whether or not the compler can elide them. If we have specific performance problems with `Base` types we can simply add some specializations.

> And we'd need to implement it case-by-case for `keytype` on the type anyway, as we can't call `keys` on the type.

This is more convincing.

**Comment by tkluck** [comment]
Reviving this pull request as I recently found myself reaching for this. Merged back master (still a clean merge) and added `valtype`.

File: base/abstractarray.jl
Comment (tkluck) [reply base/abstractarray.jl]: Not 100% sure, but a look at `SparseVector` seems to indicate that it's supposed to be `Int` even if I try my best to make it `Int16`:
```
julia> dump(SparseVector(Int16(3), Int16[2,3], [1.0,2.0]))
SparseVector{Float64,Int16}
  n: Int64 3
  nzind: Array{Int16}((2,)) Int16[2, 3]
  nzval: Array{Float64}((2,)) [1.0, 2.0]

julia> dump(eachindex(SparseVector(Int16(3), Int16[2,3], [1.0,2.0])))
Base.OneTo{Int64}
  stop: Int64 3
```
And also [the custom indices docs](https://docs.julialang.org/en/v1/devdocs/offset-arrays/) make no mention of allowing arbitrary index types. My mental model is "`Int` corresponds to `size_t`".

Comment (mbauman) [reply base/abstractarray.jl]: There are some types with non-`Int` axes but it's not a very well-trod ground, so I'd anticipate some potholes.

```julia
julia> r = typemax(Int128)-10:typemax(Int128)
170141183460469231731687303715884105717:170141183460469231731687303715884105727

julia> dump(length(r))
Int128 11

julia> dump(eachindex(r))
Base.OneTo{Int128}
  stop: Int128 11
```

Comment (tkluck) [reply base/abstractarray.jl]: Aha, thanks for finding that example @mbauman . Let me also explicitly add that
```
julia> r isa AbstractArray
true
```
(which you implied already and) which is why this is relevant.

OTOH, [the array interface](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array-1), actually does specify `Int` as the indexing type. More precisely, it makes a distinction between `IndexLinear` and `IndexCartesian` (and I should definitely add that distinction to this pull request) and prescribes a `getindex(A, i::Int)` method in the former case.

I think the sanest way forward is:

1. extend the [indexing interface](https://docs.julialang.org/en/v1/manual/interfaces/#Indexing-1) with required `keytype` and `valtype` methods
2. extend the array interface with `::Integer` and `::VarArg{<:Integer,N}` methods for `IndexLinear` and `IndexCartesian` respectively. (In the latter case, different axes may have different integer types. An interesting one is `Bool <: Integer`.)
2. do a sweep of `Base` to add these.

I'm not looking forward particularly to that last bit, but I can probably squeeze in the time if we think this is what should be merged.

Comment (mbauman) [reply base/abstractarray.jl]: I don't see this as nearly as big of a roadblock here.

1. I think `Int` and `CartesianIndex` are perfectly sane defaults that can be automatically provided for everyone.
2. You can _already_ just define your indexing methods in terms of `::Integer` without any trouble.  But if you let it, the Base fallbacks will happily convert them to `Int`s for you and call your `::Int` method.  We really don't have a demand for folks wanting to define `::Int8` indexing and finding it annoying to define both `getindex(A, I::Integer) = A[Int8(I)]` and their canonical `::Int8` method.

There's a distinction between the _implementation_ of these arrays (either `IndexLinear` or `IndexCartesian`) and their `keys`.  The former is `eachindex` (and can really be anything that's fast), whereas the latter is `keys` and defined to be either the `LinearIndices` or the `CartesianIndices` depending on the dimensionality.  And note that those two structures are (currently) completely limited to being `Int`-based.

Comment (tkluck) [reply base/abstractarray.jl]: @mbauman thanks for adding that clarity. Under this understanding, I guess the pull request is good-to-go as-is, no? Let's wait for @andyferris' opinion before resolving.

Comment (andyferris) [reply base/abstractarray.jl]: If @mbauman is OK than it's definitely OK by me!

File: base/abstractarray.jl
Comment (mbauman) [inline base/abstractarray.jl]:
```suggestion
keytype(a::AbstractArray) = keytype(typeof(a))
```

File: base/abstractarray.jl
Comment (mbauman) [inline base/abstractarray.jl]:
```suggestion
```

File: base/abstractarray.jl
Comment (mbauman) [inline base/abstractarray.jl]:
```suggestion
valtype(a::AbstractArray) = valtype(typeof(a))
```

**Review by mbauman — COMMENTED** [review]
Let's make the type-based methods the "canonical" endpoints so a custom type would only need to override that one (instead of both).

**Comment by tkluck** [comment]
All feedback is now incorporated, including the needs-* labels. The CI failures seem to be about dependency artifacts and not related to this PR.

**Comment by tkluck** [comment]
Oooh let's get this in 1.2: https://discourse.julialang.org/t/julia-1-2-feature-freeze-april-4th/22478

@mbauman or @andyferris , do either of you have commit rights? Or should we bring in someone else?

**Comment by StefanKarpinski** [comment]
It appears that this no longer needs NEWS or tests but still needs docs with compatibility annotation indicating that it was added in version 1.2. If @mbauman approves, we can merge. It would be great to get the docs with this PR but if time is an issue if you promise to add docs after the feature freeze...

**Review by mbauman — APPROVED** [review]
Thanks @tkluck! This looks great to me now. Let's merge once tests pass.

**Comment by tkluck** [comment]
@StefanKarpinski good catch, apologies. (FWIW, I (mis)understood that that label was related to the Compat.jl package, and that it would need the specific commit that introduced this feature -- i.e. that it wasn't actionable until this was merged.)

Just pushed those docs with a compatibility note.

File: base/abstractarray.jl
Comment (StefanKarpinski) [inline base/abstractarray.jl]:
```suggestion
julia> valtype(["one", "two", "three"])
```

File: base/abstractarray.jl
Comment (StefanKarpinski) [inline base/abstractarray.jl]:
```suggestion
String
```

**Comment by StefanKarpinski** [comment]
Looks great! I made a suggestion to change the `valtype` example to use a string array since it seemed more obviously different than the types of the keys of an array which could be `Int32` on some systems.

**Comment by tkluck** [comment]
Travis uncovers failing

    keytype_is_correct(Int64(1):Int64(5))

on 32-bits architecture. I'll check.

**Comment by tkluck** [comment]
buildbot is happy now, including the new test cases introduced by this pull request. Appveyor and travis complain about unrelated things and are broken on master too.

I say, good to go :)

**Review by StefanKarpinski — APPROVED** [review]
Good to go as soon as CI passes. Please squash when merging!

**Pull request merged**
